include/deal.II/base/data_out_base.h

```diff
@@ -3030,6 +3030,35 @@ class DataOutInterface
   write_pvtu_record(std::ostream &                  out,
                     const std::vector<std::string> &piece_names) const;
 
+  /**
+   * This function wraps around DataOutInterface::write_vtu() and
+   * DataOutInterface::write_pvtu_record() to avoid reimplementing the code
+   * explained in step-40. A specified @p folder and a
+   * @p filename_without_extension form the first part of the filename.
+   * The filename is then extended with the processor ID, a @p counter
+   * labeling the current cycle/timestep/etc., and finally the .vtu/.pvtu
+   * ending. Subsequently, this function calls the write_vtu(),
+   * write_pvtu_record() functions. The number of digits reserved for the
+   * processor ID in the filename is determined from the maximum number of
+   * MPI ranks, determined from a specified @p mpi_communicator, for which
+   * the default value is MPI_COMM_WORLD. Since the number of output files to be
+   * written depends on the application, the number of digits to be reserved in
+   * the filename can be specified as parameter @p n_digits_counter, and the number
+   * is not padded with leading zeros if this parameter is left at its default
+   * value numbers::invalid_unsigned_int. Note that only one processor needs to
+   * generate the .pvtu file, where processor zero is chosen to take over this
+   * job.
+   *
+   * @author Niklas Fehn, Martin Kronbichler, 2019
+   */
+  void
+  write_vtu_with_pvtu_record(
+    const std::string &folder,
+    const std::string &filename_without_extension,
```

Comment:
I think we can reduce redundancy by extracting the filename without folders from the string. On Unix-like systems, one can search whether `filename_without_extension` contains any `/` character and remove everything up to that one:
```c++
std::string filename_without_folder;
if (filename_without_extension.find('/') != std::string::npos)
  filename_without_folder = std::string(filename_without_extension.begin()+
                                        filename_without_extension.find_last_of('/')+1,
                                        filename_without_extension.end());
else
  filename_without_folder = filename_without_extension;
Assert(!filename_without_folder.empty(), ExcMessage("Cannot have empty file name"));
```
On Windows, that character would be `\`. Probably there is even a boost function for that, I have not checked.

Reply:
In my opinion, this has some disadvantages and I want to share them before deciding for this variant. For me, ``folder`` + ``filename`` is more natural. On the one hand, the code has to separate into folder and filename anyway, so the implementation suggested above is more complicated and error prone. On the other hand, I think the suggested version is not self-documenting, as it is not clear what ``filename_without_extension`` is. Sure, one might define ``filename`` in a way that it always includes the folder. However, assume we have two parameters ``folder`` and ``filename``, I think this does not allow misunderstandings, as it is clear that one does not expect that ``filename`` also includes the folder in this case. Finally, in my experience an application code has separate parameters for folder and filename anyway as soon as it becomes more complicated (because the user has to specify several names for different output files in an input file that collects all parameters but does not want to rewrite the folder for all instances, and hence uses a separate parameter for the folder). I have the concern that in practice we first combine folder and filename while the code later separates them again, while it is the aim of this PR to simplify interfaces. All tutorials either write the output in the current working directory or use two separate parameters for folder and filename (see step-42). Hence, there is no compatibility problem in the other direction in the sense that the filename has to be separated into folder and filename manually by the user before calling the new function ``write_vtu_with_pvtu_record()``.

Reply:
I think it would make more sense to write a free function that can separate a string into folder and filename_without_folder in case that some code uses this format (a single filename including the folder). It needs several lines of code to implement this and why should this very general functionality be implemented in ``write_vtu_with_pvtu_record``?

Reply:
I fully agree on the case that larger codes will put this into subfolders - my argument was that it is an optional parameter for simple programs, setting the first argument of this function to `""`.
On the other hand, I agree that `filename_without_extension` would be ambiguous in case it should contain a folder name, but that is just a matter of naming. Funny though that I initially thought your interface @nfehn makes it more easy to make small mistakes, so here is something more to be documented: when you have separate `folder` and `filename` strings the `folder` argument must include the trailing "/" character - not a big deal but easily forgotten when you are used to bash that does the magic for you.

> I have the concern that in practice we first combine folder and filename while the code later separates them again, while it is the aim of this PR to simplify interfaces.

But what I suggested would just make the internal implementation more complicated, not the interface.

In any case, I do not feel strongly about either way, as long as it is well documented what is required.